### ISSUE TITLE
Http2 buffer allocation reduction

### ISSUE THREAD

**Issue by Scottmitch** [netty/netty#3390]
Based off of @louiscryan's https://github.com/netty/netty/pull/3385#discussion_r24048271

Can we reduce the amount of buffers that must be allocated to write a FRAME to the wire?  Some methods of the `Http2FrameWriter` interface take a`ByteBuf` objects as a parameter, and then allocate at least 1 more buffer to write the headers and remaining payload.  This may have benefits for frame types which are expected to be sent in high volumes.  For example the `writeData` method may be a good candidate for this investigation.

**Comment by Scottmitch** [comment]
@louiscryan @nmittler FYI.

**Comment by Scottmitch** [comment]
Something to consider.  Some frames (DATA) for example have the data in the middle of the user's data.  For example DATA FRAME:

1. Header
2. User Data
3. Padding

So the allocator has to take this into account and write the header before returning the buffer for the user to write to.

**Comment by nmittler** [comment]
@Scottmitch @louiscryan are we waiting on anything before starting work on this?   If not, I can take a crack.

**Comment by louiscryan** [comment]
Nothing comes to mind though but it would be interesting to see the impact
of Scotts current improvements on his new benchmark

**Comment by Scottmitch** [comment]
Although no interface changes were done in PR https://github.com/netty/netty/pull/3385 it may be worth taking these results into account.  The padding issue may be mitigated with https://github.com/netty/netty/pull/3385.

**Comment by nmittler** [comment]
@Scottmitch is #3385 ready to cherry-pick?

**Comment by Scottmitch** [comment]
@nmittler I was going to do performance testing vs master branch to be sure it is an improvement. However I think you and @louiscryan have done great on the reviews...I think it will be at least as good if not a bit better than the master.  If you are satisfied I don't want to hold you back. I'm traveling again this weekend so feel free to cherry pick if you are satisfied.

**Comment by normanmaurer** [comment]
@Scottmitch @nmittler @louiscryan +1

**Comment by louiscryan** [comment]
SGTM. We can (and should) do a before-vs-after later once we've nailed down
the benchmarking PR

**Comment by nmittler** [comment]
@Scottmitch FYI, #3385 has been cherry-picked.

**Comment by nmittler** [comment]
Folks, I started hacking away locally to see what this might look like ... nothing really in a state yet that's worth posting, but it's sparked some food for thought...

The general idea would be to change `Http2FrameWriter` to have factory methods for frame types that have a payload (e.g. data, ping, goaway).  The caller would then populate the payload portion of the returned `Frame` before writing it.

For frames sent before flow control windows kick in, we could combine the frame header and payload into a single buffer.  There would still be a second call to `ctx.write()` for the padding (if sending padding), but no alloc since we're now using slices from a global padding buffer.

Once flow control kicks in, we start splitting DATA frames.  In this case I think we could still avoid allocating new buffers and keep each send to at most 2 calls to `ctx.write()`.  When called to create a new DATA frame, we would allocate a buffer to contain the header and the entire payload (minus the padding since that's sent separately). The user would then fill out the payload and write the frame.  This would trigger the call to the flow controller.  As the flow control writes portions of the frame, we would update and move the DATA frame header in the buffer, overwriting parts of the data that have already been sent.

With all of that said, this all would be a fairly significant change and it might be better to wait for @Scottmitch finish his work with the benchmarks to see what the performance advantage is (if any) for using fewer buffers.

Thoughts?

**Comment by Scottmitch** [comment]
@nmittler - Thanks for the description!  I agree that the main opportunity is to reduce an additional allocation for each frame with a payload (due to the HEADERS).  Can you clarify the following?

> As the flow control writes portions of the frame, we would update and move the DATA frame header in the buffer, overwriting parts of the data that have already been sent.

Why do we need to move the header around?

> With all of that said, this all would be a fairly significant change and it might be better to wait for @Scottmitch finish his work with the benchmarks to see what the performance advantage is (if any) for using fewer buffers.

I agree that before spending the effort (and potentially adding complexity) it would be useful to know what kind of returns to expect.  I started a new job this week and so I am working through the approval process for committing back to Netty.  It may be a bit before PR #3392 is cherry-picked but we can collect some preliminary results with what is there now to get a ballpark.

**Comment by nmittler** [comment]
@Scottmitch ignore my comment about moving the header around ... that wouldn't work.  I think we can only combine the frame header and data into a single buffer on the first write.  Any further splitting of the data will require an additional buffer and write operation for the header.

**Comment by Scottmitch** [comment]
@nmittler I don't understand the following:

> Any further splitting of the data will require an additional buffer and write operation for the header.

Why do we have to write the header differently when splitting occurs or need an additional buffer?

Let me lay out my mental model and get your thoughts:

1. User asks for the first payload buffer for a stream for frame type `x` to be allocated from our allocator (lets assume payload size is supplied).
2. We allocate a buffer `payload size + frame type x header size`.  If we know the header now then write the header in.  If we don't know the header now then set writer index to `frame type x header size`.
3. User calls `FrameWriter.writeXXX(...)`.
4. In `writeXXX(...)` we can check if this is the first write for this stream (using a decorated bytebuf returned by our allocator, stream state checking, etc...).  If it is the first write and we have not populated the header yet then write the header into the buffer and then write the buffer to the channel. Otherwise just write the buffer to the channel.

If a size is not supplied then we can allocate for at least as much size as the header frame type (plus some additional?).  If the user wants/needs to allocate multiple buffers (i.e. data is generated asynchronously) then is there a need for any other buffers besides the first to be treated differently?

On a separate note (another issue perhaps)...we are partially consuming buffers for the DATA frame type and we may want to consider supporting an interface for partial buffer "discarding" (i.e. hooks to call [discardSomeReadBytes](https://github.com/netty/netty/blob/master/buffer/src/main/java/io/netty/buffer/ByteBuf.java#L476) or some other user supplied discard routine)

**Comment by nmittler** [comment]
@Scottmitch I think we might be talking about slightly different things.  Let me see if I can clarify what I was saying.  The problem I was trying to address was that of writing DATA frames, since they are the only frames that are flow controlled and can thus be split after the initial allocation.

## New Frame Construction Pattern
My strategy for this optimization was going to change the encoder (as well as FrameWriter) to be a factory for DATA frame objects which expose a writable slice for the data portion.  Something like this:

    interface Http2FrameWriter {
        ....
        interface Frame {
            ByteBuf payload();
            void writeTo(ctx, promise);
        }
        ...
        Frame newDataFrame(...);
    }

The idea here is that the call to `newDataFrame` will internally allocate a single `ByteBuf` of size `payload size + header size` and the `Frame` will then expose a writable slice of that buffer for writing the `payload`.

Since the encoder also implements this interface, it would expose the same construction pattern to the application for DATA frames.  The application would create a new DATA frame, fill in the payload portion, and then call the `writeTo()` method, which would trigger the encoder to pass it on to flow control. The application of flow control is where this optimization starts to break down.

If flow control is able to send it all immediately, there is no problem - we just write the entire frame buffer and we're done.  However if flow control has to split the frame, we have some work to do.

## Splitting DATA frames
Once we've determined that the frame must be split, at a minimum we have to update the DATA frame header for the first write.  If padding > 0, that means that the header may have to shrink as well (currently we only send padding at the very end after all data has been sent - we probably want to change that eventually). Since the frame header is initially in the same `ByteBuf` as the DATA, for the very first write we could just manipulate the frame header in-place before the write occurs. We create a new slice containing this updated header and the portion of the DATA that we want.  This can all be sent with a single `write()` (if we eventually start splitting padding as well, there would be a second `write()` for a portion of the padding).

After the first write, however, we can't add the header for the next frame in the same `ByteBuf`, since this would corrupt the previously sent frame if the write operation hadn't yet completed.  In this case, we have to allocate a new `ByteBuf` for the frame header.  So all subsequent `write()` operations for a split frame will require a minimum of 2 writes: frame header, chunk of the DATA, and optionally padding.

## Thoughts
So with all of that said, this optimization would be most effective if flow control is not splitting the frame at all, or if the number of splits per send is very low.  However, if we assume that over time flow control will be splitting frames fairly regularly this optimization will have diminishing returns.

Does that clarify my previous comments at all?  Do you have any other suggestions for how we might tackle this?

**Comment by Scottmitch** [comment]
@nmittler - Yes thanks for clarifying, and I agree with your diminishing returns under load analysis.  I was leaving out the part where the frame header for frames after the "split" (duh :) ).

You description on allocations and writes seems reasonable, but let me reiterate/summarize.

Initial Frame (before split condition is known):
- 1 allocation 2 writes (header,data + padding).  Space for header is included in the allocation, and the write operation can fill in the header details.

Each frames after split:
- 1 additional allocation (header) 3 writes (header + data + padding).

So as you pointed out this optimization buys us 1 allocation and 1 write savings for the initial frame.  For non flow controlled frames (with a payload) we will always get this benefit.

**Comment by nmittler** [comment]
@Scottmitch yes, exactly.  So if our analysis of the benchmarks shows that removing the extra buffer allocation for the frame header is a possible win, this may still be something to consider.  Otherwise it's probably not worth doing.

**Comment by Scottmitch** [comment]
+1.

**Comment by nmittler** [comment]
@Scottmitch where do we currently stand with this? Are we satisfied with the current performance? Should we close this?

**Comment by Scottmitch** [comment]
@nmittler - You guys are doing more production benchmarking and if you are satisfied that make me satisfied :)  We can re-visit later if necessary.
